File: src/main/java/com/cfgglobal/test/domain/Branch.java
package com.cfgglobal.test.domain;


import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.FieldDefaults;
import org.hibernate.annotations.DynamicInsert;
import org.hibernate.annotations.DynamicUpdate;

import javax.persistence.*;
import java.util.List;

@Entity
@Getter
@Setter
@DynamicUpdate
@DynamicInsert
@FieldDefaults(level = AccessLevel.PRIVATE)
public class Branch extends BaseEntity {


    String name;

    @ManyToOne(cascade = {CascadeType.REFRESH, CascadeType.REMOVE}, optional = true)
    @JoinColumn(name = "parent_id")
    Branch parent;
    @OneToMany(cascade = {CascadeType.REFRESH, CascadeType.REMOVE}, fetch = FetchType.LAZY, mappedBy = "parent")
    List<Branch> children;
/*
    @ManyToMany(cascade = CascadeType.REFRESH)

     List<User> users;*/

}

File: src/main/kotlin/com/cfgglobal/test/domain/Branch.kt
package com.cfgglobal.test.domain


import com.cfgglobal.generator.metadata.FieldFeature
import org.hibernate.annotations.DynamicInsert
import org.hibernate.annotations.DynamicUpdate
import org.hibernate.validator.constraints.Length
import javax.persistence.*
import javax.validation.constraints.NotNull

@Entity
@DynamicUpdate
@DynamicInsert
data class Branch(

        @Id
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        var pk: Long? = null,

        @NotNull
        @Column(length = 10)
        val name: String? = null,

        @ManyToOne(cascade = [(CascadeType.REFRESH), (CascadeType.REMOVE)], optional = true)
        @JoinColumn(name = "parent_id")
        val parent: Branch? = null,
        @OneToMany(cascade = [(CascadeType.REFRESH), (CascadeType.REMOVE)], fetch = FetchType.LAZY, mappedBy = "parent")
        val children: MutableList<Branch> = mutableListOf()

) : BaseEntity()
